ui/src/reusable_ui/components/dropdowns/Dropdown.tsx

```diff
@@ -0,0 +1,174 @@
+
+import {ClickOutside} from 'src/shared/components/ClickOutside'
+import {
+  ComponentColor,
+  ComponentSize,
+  IconFont,
+  DropdownMenuColor,
+} from 'src/reusable_ui/types'
+import DropdownButton from 'src/reusable_ui/components/dropdowns/DropdownButton'
+import FancyScrollbar from 'src/shared/components/FancyScrollbar'
+import {ErrorHandling} from 'src/shared/decorators/errors'
+import './Dropdown.scss'
+
+interface Props {
+  children: Array<JSX.Element | JSX.Element[]>
+  onChange: (value: any) => void
+  selectedItem: string
+  color?: ComponentColor
+  menuColor?: DropdownMenuColor
+  size?: ComponentSize
+  disabled?: boolean
+  width?: number
+  icon?: IconFont
+  wrapText?: boolean
+  customClass?: string
+  maxMenuHeight?: number
+}
+
+interface State {
+  expanded: boolean
+}
+
+@ErrorHandling
+class Dropdown extends Component<Props, State> {
+  public static defaultProps = {
+    color: ComponentColor.Default,
+    size: ComponentSize.Small,
+    disabled: false,
+    width: 120,
+    wrapText: false,
+    maxMenuHeight: 250,
+    menuColor: DropdownMenuColor.Sapphire,
+  }
+
+  constructor(props: Props) {
+    super(props)
+
+    this.state = {
+      expanded: false,
+    }
+  }
+
+  public render() {
+    const width = `${this.props.width}px`
+
+    return (
+      <ClickOutside onClickOutside={this.collapseMenu}>
+        <div className={this.containerClassName} style={{width}}>
+          {this.button}
+          {this.menu}
+        </div>
+      </ClickOutside>
+    )
+  }
+
+  private toggleMenu = (): void => {
+    this.setState({expanded: !this.state.expanded})
+  }
+
+  private collapseMenu = (): void => {
+    this.setState({expanded: false})
+  }
+
+  private get containerClassName(): string {
+    const {color, size, disabled, wrapText, customClass} = this.props
+
+    return classnames(`dropdown dropdown-${size} dropdown-${color}`, {
+      disabled,
+      'dropdown-wrap': wrapText,
+      [customClass]: customClass,
+    })
+  }
+
+  private get button(): JSX.Element {
+    const {selectedItem, disabled, color, size, icon} = this.props
+    const {expanded} = this.state
+
+    return (
+      <DropdownButton
+        label={selectedItem}
+        active={expanded}
+        color={color}
+        size={size}
+        icon={icon}
+        disabled={disabled}
+        onClick={this.toggleMenu}
+      />
+    )
+  }
+
+  private get menu(): JSX.Element {
+    const {selectedItem, maxMenuHeight, menuColor} = this.props
+    const {expanded} = this.state
+
+    if (expanded) {
+      return (
+        <div
+          className={`dropdown--menu-container dropdown--${menuColor}`}
+          style={this.menuStyle}
+        >
+          <FancyScrollbar
+            autoHide={false}
+            autoHeight={true}
+            maxHeight={maxMenuHeight}
+          >
+            <div className="dropdown--menu">
+              {this.flatChildren.map((child: JSX.Element) =>
+                React.cloneElement(child, {
```

Comment:
I updated the key to not use `uuid` so it won't re-render so much. Not sure if there's a cleaner way to handle the double-mapping process I employed during flattening

Reply:
We might be able to avoid cloning via component specialization. Something like:

```
// ...
<div className="dropdown-menu">
  {this.props.children.map(child => {
    if (child instanceof Dropdown.Item) {
      return (
        <Dropdown.Item
          {...child.props}
          key: `dropdown-menu--${child.props.text}`,
          selected: child.props.text === selectedItem,
          onClick: this.handleItemClick
        />
      )
    } else {
      return child
    }
  })}
</div>
```

Then hopefully this would let us avoid flattening fragments out of children, too.